DIPs/DIP27.md

```diff
@@ -0,0 +1,145 @@
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP proposes to radically simplify functions' behavior first class functions. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+D currently has first class function, and what I'll call second class function. They are defined as follow :
+
+```d
+ReturnType secondClass(Parameters) {
+  // Function body.
+}
+
+enum firstClass = function ReturnType(Parameters) {
+  // Function body.
+};
+```
+
+First class and second class function both have their own behavior. In this DIP, I propose to define second class function such as :
+ * Taking the address of a second class function return a first class function.
+ * A second class function implicitely convert to a first class function.
+
+The behavior of a first class function is altered such as it is possible to use optional parentheses. See the optional parentheses section below for the details of the behavior.
+
+The function keyword in is expression now match function type instead of second class function. See the is expression section below for detail behavior.
+
+### Rationale
+
+D is a very complex language. It promotes several paradigms while being a system language. If this is a force of D, this is also a source of trouble. In fact every feature can combine itself why many other features, some of them from different paradigms - other way of thinking - and every single special case, cause a combinatorial explosion of special cases when other features are involved.
+
+This problem is experienced by many D users who try to use advanced features together in non trivial projects. Combined together, special cases lead to surprising behaviors, unnecessary blockades, and sometime compiler bugs.
+
+If it is impairing users, it is also impairing polishing the language, as many corner cases have to be considered, and cause difficulties for compiler implementers and libraries creators that uses generic techniques.
+
+To solve, or at least reduce that problem, D specification needs to promote simplicity (as opposed to complexity, not difficulty). To move toward this goal, this DIP reduce all function to a single entity : the D first class function. It rid of functions like defined in C or C++, as they are useless and essentially accidental complexity.
+
+### Optional parentheses
+
+Redundant parenthesis can be a burden for argument-less functions (or single argument function calls used as UFCS). An implicit function call is performed in the following cases :
+
+1. When .identifier lookup fails on the function :
+
+```d
+uint foo() {
+  return 0;
+}
+
+void bar(uint function() a) { writeln("function bar"); }
+void bar(uint a) { writeln("uint bar"); }
+
+foo.bar(); // function bar
```

Comment:
Also as to motivation, even phobos devs seems to not get it right (and you need at least 2 of them to be fooled for something to be merged). It is a dead giveaway that there is  problem to fix.